fpga_interchange/site_lut_mapping_cache.cc

```diff
@@ -0,0 +1,187 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2021  Symbiflow Authors
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "nextpnr.h"
+#include "archdefs.h"
+#include "site_arch.h"
+#include "site_lut_mapping_cache.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+// ============================================================================
+
+SiteLutMappingKey SiteLutMappingKey::create (const SiteInformation& siteInfo) {
+    const Context *ctx = siteInfo.ctx;
+
+    // Look for LUT cells in the site
+    std::vector<CellInfo*> lutCells;
+    lutCells.reserve(siteInfo.cells_in_site.size());
+
+    for (CellInfo* cellInfo : siteInfo.cells_in_site) {
+
+        // Not a LUT cell
+        if (cellInfo->lut_cell.pins.empty()) {
+            continue;
+        }
+
+        // Not bound to a LUT BEL
+        BelId bel = cellInfo->bel;
+        const auto &bel_data = bel_info(ctx->chip_info, bel);
+        if (bel_data.lut_element == -1) {
+            continue;
+        }
+
+        lutCells.push_back(cellInfo);
+    }
+
+    // Sort cells by BEL indices to maintain always the same order
+    std::sort(lutCells.begin(), lutCells.end(),
+              [](const CellInfo* a, const CellInfo* b)
+    {
+        return a->bel.index > b->bel.index;
+    });
+
+    // Initialize the key
+    SiteLutMappingKey key;
+    key.tileType = siteInfo.tile_type;
+    key.siteType = ctx->chip_info->sites[siteInfo.site].site_type;
+    key.numCells = 0;
+
+    // Get bound nets. Store localized (to the LUT cluster) net indices only
+    // to get always the same key for the same LUT port configuration even
+    // when the actual global net names are different.
+    dict<IdString, int32_t> netMap;
+    for (CellInfo* cellInfo : lutCells) {
+
+        NPNR_ASSERT(key.numCells < SiteLutMappingKey::MAX_LUT_CELLS);
+        auto& cell = key.cells[key.numCells++];
+
+        cell.type       = cellInfo->type;
+        cell.belIndex   = cellInfo->bel.index;
+
+        memset((void*)cell.conns, 0,
```

Comment:
once this becomes `std::array`, this could use `std::fill` instead

Reply:
allright, I'll change those to `std:array`.